src/servers/Http1Server.cc

```diff
@@ -189,10 +189,13 @@ Http::One::Server::buildHttpRequest(Http::StreamPointer &context)
     // some code still uses Host directly so normalize it using the previously
     // sanitized URL authority value.
     // For now preserve the case where Host is completely absent. That matters.
-    if (const auto x = request->header.delById(Http::HOST)) {
-        debugs(33, 5, "normalize " << x << " Host header using " << request->url.authority());
-        SBuf tmp(request->url.authority());
-        request->header.putStr(Http::HOST, tmp.c_str());
+    if(request->header.findEntry(Http::HOST)){
+        SBuf domain(request->url.authority());
+        HttpHeaderEntry* host = new HttpHeaderEntry(Http::HOST, NULL, domain.c_str());
+        if(int x = request->header.delById(Http::HOST)){
+            debugs(33, 5, "normalize " << x << " Host header using " << request->url.authority());
+            request->header.insertEntry(host);
```

Comment:
Please add a reference to the RFC requirement about these.
```suggestion
            // RFC 9110 section 7.2: agent that sends Host SHOULD send it as the first field in the header section
            request->header.insertEntry(host);
```

Reply:
In regards to a specific case, 'linuxconfig.org' for example is blocking the request coming from Squid that is being used as a cache server in my situation. 

I became curious and inspected which the only noted difference between raw browser request and going through Squid was the Host header position in the header list, made the submitted adjustment, and it seems to have done the trick. 

Another thing keeping the Host header 'on top' seems to overcome is getting CAPTCHA by sites such as 'linuxquestions.org' for example as well as a handful of others.

Hard to believe that simply the Host header position is triggering such things, but so far, it seems to be exactly the issue unless I'm tripping.

Regarding not updating the Host header if not needed, agreed, leave it be, but in regards to updating in place the thing that crossed my mind is along the lines of memory addressing/allocation becoming jacked. Not sure as need to learn more about Squid code, just came up with this easy fix which seems to overcome the above stated issue which can be a real nuisance sometimes aside from the fact that Squid seems to do an awesome job as a cache server otherwise. 

I will admit that don't totally understand the reason for the Host header being mangled in the first place?

Reply:
If it is being mangled here simply for sanitation purposes, then seems like it could be done way before this point which would ultimately eliminate the need for the code section here in the first place.

Personally need to dig further into Squid code as only took a few hours to track this down after getting tired of being blocked/checked by some sites which seems an issue that could deter others from wanting use it if coming across the same issue. 

Again, so far, other than this issue, Squid seems to do a rocking job as a cache server.

Reply:
There are two reasons for mangling the Host header.

1) to ensure that it is valid when sent. If the client presented an invalid header, or if Squid internally changed the destination URL. The latter may occur at any point up to and including peer selection (cache_peer forceddomain option).

2) to avoid old and buggy squid code. This is what the existing comment refers to I am not sure if those bugs still exist.

It is mangled here because this function is where the outbound HTTP headers are initially created. Any previous code would be altering Squid's view of the clients request, which is semantically supposed to be a different message and needs preserving for logging later.

Reply:
@yadij, I added that (important) RFC quote to the PR description (for now). I do not think we should add it to the code itself for several reasons, in no particular order:

1. Even if this code enforces (or preserves) the correct Host header position, Squid often inserts another header before Host. We have to do more to actually enforce that RFC SHOULD.
2. I doubt it is a good idea to move the Host header to the correct position. If some bot is attacking a CDN while hiding behind Squid, and that CDN can fingerprint that bot by that bad Host header position, let them block bot+Squid combo. The opposite "let's fix what we can" strategy will only result in them fingerprinting and blocking _Squid_ instead (in all environments) -- a far worse outcome! As you know, these little "fixes" sometimes misfire for unexpected peer compatibility reasons, requiring configuration options or additional compensating hacks. I think preserving the user agent Host header position, good or bad, is an overall better strategy, especially since that position is correct in the vast majority of cases.
3. insertEntry() is very expensive performance-wise. Its push_front() behavior is a surprise to many developers. It is used incorrectly by some code. We should work towards removing it completely.

If you insist, we could add a source code comment with that RFC quote and augment it with similar caveats/explanations.